demo_fastapi/sales.py

```diff
@@ -57,27 +73,118 @@ def create_order(
     )
+
+    # save to the database
+    cursor = cnxn.cursor()
+    try:
+        cursor.execute(
+            "INSERT INTO ShoppingList VALUES (?,?,?,?,?)",
+            datestamp,
+            buyer,
+            apples,
+            oranges,
+            order_id,
+        )
+        cursor.commit()
+    except pyodbc.DatabaseError as err:
+        raise HTTPException(
+            status_code=500,
+            detail="Error updating database! Recheck your entries.\n" + str(err),
+        )
+
     return order
 
 
-# @app.put("/orders/{order_id}")
-# def update_order(
-#     order_id: int,
-#     datestamp: str,
-#     buyer: str,
-#     apples: PositiveInt = None,
-#     oranges: PositiveInt = None,
-# ):
-#     # update order with ID "order_id"
-#     order = Order(datestamp=datestamp, buyer=buyer, apples=apples, oranges=oranges)
-#     return {"order_id": order_id, "order": order}
+@app.put("/orders/")
+def update_order(
+    order_id: int,
+    datestamp: str = None,
+    buyer: str = None,
+    apples: PositiveInt = None,
+    oranges: PositiveInt = None,
+):
+    """update order with ID = order_id"""
+    cnxn = app.state.connection_manager.connection
+    cursor = cnxn.cursor()
+    cursor.execute("SELECT * FROM ShoppingList WHERE id = ?", order_id)
+    row = cursor.fetchone()
 
+    datestamp = datestamp if datestamp is not None else row.datestamp
+    buyer = buyer if buyer is not None else row.buyer
+    apples = apples if apples is not None else row.apples
+    oranges = oranges if oranges is not None else row.oranges
 
-# @app.get("/orders/{order_id}")
-# def read_order(order_id: int):
-#     # get data of order with ID "order_id"
-#     return {"item_id": order_id, "order": order_id}
+    try:
+        cursor.execute(
+            "UPDATE ShoppingList SET datestamp = ?, buyer = ?, apples = ?, oranges = "
+            "? WHERE id = ?",
+            datestamp,
+            buyer,
+            apples,
+            oranges,
+            order_id,
+        )
+        cursor.commit()
+    except pyodbc.DatabaseError as err:
+        raise HTTPException(
+            status_code=418,
+            detail="Error updating database! Recheck your entries.\n" + str(err),
+        )
+
+    # order = Order(datestamp=datestamp, buyer=buyer, apples=apples, oranges=oranges)
+    return Order(
+        id=order_id, datestamp=datestamp, buyer=buyer, apples=apples, oranges=oranges
+    )
+
+
+@app.get("/orders/")
+def read_order(order_id: int):
+    """Read the order with ID = order_id from the database and print it in the app"""
+    cnxn = app.state.connection_manager.connection
+    cursor = cnxn.cursor()
+    try:
+        cursor.execute("SELECT * FROM ShoppingList WHERE id = ?", order_id)
+        # cursor.commit()
+    except pyodbc.DatabaseError as err:
+        raise HTTPException(
+            status_code=418,
+            detail="Error reading database! Recheck your entries.\n" + str(err),
```

Comment:
same comment as above where the same error detail appears :-)

Reply:
When the order_id is unknown, we say: Hey I dont know this id!

Reply:
3 Errors: 
- order id unknown 
- database connection broken
- table does not exist